[InstCombine] Fold (X / C) < X and (X >> C) < X into X > 0

Poseydon42 opened the pull request:
Proofs:
https://alive2.llvm.org/ce/z/52droC
This resolves #85313

goldsteinn commented:
Just a note on the proofs.
You really only proved for the exact constant chosen i.e `lshr x, 1` in the case of the `lshr` proofs. That doesn't necessarily transfer to all other constants (in fact 0 is incorrect).
See some general proofs: https://alive2.llvm.org/ce/z/3YWXgD
I filled in the `sdiv` cases, can you please complete the `udiv`/`lshr`/`ashr` ones.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7406,6 +7406,31 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
+              m_CombineOr(m_SDiv(m_Value(Dividend), m_ImmConstant(Divisor)),
+                          m_UDiv(m_Value(Dividend), m_ImmConstant(Divisor)))) &&
+        match(Op1, m_Deferred(Dividend)) &&
+        !(Divisor->isZeroValue() || Divisor->isOneValue())) {
```
Divisor->ugt(1)

goldsteinn replied:
This is buggy I think. Imagine a non-splat vec with denom `<i8 0, i8 123>`. Not zero/ones but still incorrect. Think you should change `m_ImmConstant` to `m_APInt` then just use `ugt(1)`

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7406,6 +7406,31 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
+                          m_UDiv(m_Value(Dividend), m_ImmConstant(Divisor)))) &&
+        match(Op1, m_Deferred(Dividend)) &&
+        !(Divisor->isZeroValue() || Divisor->isOneValue())) {
+      return new ICmpInst(ICmpInst::ICMP_SGT, Dividend,
```
You are missing a check that in the incoming Predicate was `ICMP_SLT`.

In general though, can you handle a lot more predicates and the general pattern `ICmpInst::getSwappedPredicate(I.getPredicate())`

goldsteinn replied:
Also note you can handle just about any predicate here.
For `sdiv`/`ashr`, you need `!ICmpInst::isUnsigned(I.getPredicate())`. For `udiv`/`lshr`, any predicate works.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7406,6 +7406,31 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
+    Constant *Shift;
+    Value *V;
+    if (match(Op0, m_LShr(m_Value(V), m_ImmConstant(Shift))) &&
+        match(Op1, m_Deferred(V)) && !Shift->isZeroValue()) {
```
`M_Deferred` isn't really right here. Thats for if you are matching a specific `V` set earlier in the same `match` statement.
Here you want to use `m_Specific`.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7406,6 +7406,31 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
+                          Constant::getNullValue(V->getType()));
+    }
+  }
+
```
This code should probably all be in the `foldICmpCommutative` function.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7406,6 +7406,31 @@ Instruction *InstCombinerImpl::visitICmpInst(ICmpInst &I) {
+    if (match(Op0,
+              m_CombineOr(m_SDiv(m_Value(Dividend), m_ImmConstant(Divisor)),
+                          m_UDiv(m_Value(Dividend), m_ImmConstant(Divisor)))) &&
+        match(Op1, m_Deferred(Dividend)) &&
```
Really this can just `Op1 == Dividend` (likewise below in the `lshr` case).

goldsteinn commented:
Also welcome and thank you for the commit. There a rough draft of a contribution guide for `InstCombine` here: #79007

You might find it useful.

Poseydon42 commented:
I have modified the code to work with all predicates for `udiv`/`lshr` and unsigned predicates for `sdiv`, however I believe that `ashr` does not work with this type of fold. As you can see [here](https://alive2.llvm.org/ce/z/cVMB9e), the only two predicates that it does work with are `slt` and `sge`. Should I include these two folds for `ashr` or should I ignore it altogether and only focus on the other 3 instructions?

goldsteinn commented:
> I have modified the code to work with all predicates for `udiv`/`lshr` and unsigned predicates for `sdiv`, however I believe that `ashr` does not work with this type of fold. As you can see [here](https://alive2.llvm.org/ce/z/cVMB9e), the only two predicates that it does work with are `slt` and `sge`. Should I include these two folds for `ashr` or should I ignore it altogether and only focus on the other 3 instructions?

Think made a mistake when merging your updates.

Poseydon42 commented:
Yeah, that merge was an accident, I'll remove it before pushing the most recent changes. It seems to me like handling `ashr` would just unnecessarily increase the complexity of the code without providing significant benefit, so I'll omit it. The only question I have now is whether I should add tests for all three instructions I'm working with and all 10 (or 6) comparison predicates, making it 26 tests in total, which, together with negative tests, will bring it closer to 30, or should I simply test 2-3 predicates for each of the three instructions?

goldsteinn commented:
> Yeah, that merge was an accident, I'll remove it before pushing the most recent changes. It seems to me like handling `ashr` would just unnecessarily increase the complexity of the code without providing significant benefit, so I'll omit it. The only question I have now is whether I should add tests for all three instructions I'm working with and all 10 (or 6) comparison predicates, making it 26 tests in total, which, together with negative tests, will bring it closer to 30, or should I simply test 2-3 predicates for each of the three instructions?

So generally you should be testing different control flows moreso than code.

In this case you have all predicates have the same control flow for `lshr`/`udiv`, so maybe between the two test one equality pred, one signed, and one unsigned (i.e eq, sle, ugt). Then a negative test for each maybe with an out of bounds constant.

For `sdiv` maybe ne / sgt with two negative tests  (out of bounds/unsigned pred).

Then you can pretty randomly make 1-2 of the tests use vec.

Please be sure to have a negative test with a vec where one vec element is invalid and the other is valid.

Poseydon42 committed:
[InstCombine] Add test cases for folds (X / C) cmp X and (X >> C) cmp X

Poseydon42 commented:
Updated proofs (26 in total) can be found [here](https://alive2.llvm.org/ce/z/TDRxuG). I've fixed most of the issues mentioned here, the only two things I'm worried about are:
  1.The current code does not work with vector constants that have different values in each component (e.g. take a look at test case `lshr_by_const_cmp_sle_value`). I believe this is caused by `m_APInt` not matching when a vector constant has different components, but I am unsure about that. Can something be done about this or should I leave this particular scenario unaffected by the optimization?
  2. It appears that `lshr` already has a similar fold implemented, since test case `lshr_by_const_cmp_ule_value` was not affected by the changes that I've made (it folded to `ret i1 true` both before and after the changes to the code). Can someone confirm that this is the case, and if so, should I remove the `lshr` code path from my PR?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -11,6 +11,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "InstCombineInternal.h"
+#include "llvm/ADT/APInt.h"
```
This shouldn't be needed.

goldsteinn replied:
Likewise below.

goldsteinn commented:
> Updated proofs (26 in total) can be found [here](https://alive2.llvm.org/ce/z/TDRxuG). I've fixed most of the issues mentioned here, the only two things I'm worried about are: 1.The current code does not work with vector constants that have different values in each component (e.g. take a look at test case `lshr_by_const_cmp_sle_value`). I believe this is caused by `m_APInt` not matching when a vector constant has different components, but I am unsure about that. Can something be done about this or should I leave this particular scenario unaffected by the optimization? 

Yes that is one of the drawbacks of APInt, it only match if the vector is a splat (all elements that same).
You could use: `m_SpecificInt_ICMP` (see PatternMatch.h).
So for the div cases: `m_SpecificInt_ICMP(ICmpInst::ICMP_UGT, APInt::getOneBitSet(Op0->getType()->getScalarSizeInBits(), 1))` (not the cleanest API to use).

> 2. It appears that `lshr` already has a similar fold implemented, since test case `lshr_by_const_cmp_ule_value` was not affected by the changes that I've made (it folded to `ret i1 true` both before and after the changes to the code). Can someone confirm that this is the case, and if so, should I remove the `lshr` code path from my PR?

See: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/InstructionSimplify.cpp#L3172
No need to go out of your way to avoid that case. Its not extra code on your end to handle (and its correct). I wouldn't use it as a test, however, as its not going through your code path.

Poseydon42 committed:
[InstCombine] Fold (X / C pred X) and (X >> C pred X) into X ~pred 0

Poseydon42 commented:
Right, after a painful hour of rewriting git history multiple times, it seems to me that I've covered everything that was mentioned here before. Please let me know if there's anything else for me to fix or improve.

Poseydon42 commented:
Oops, looks like I forgot to run tests on my machine before pushing the changes. I've localised the source of one of many failing tests to this function in `getelementptr.ll`:
```
define i1 @test10(ptr %x, ptr %y) {
; CHECK-LABEL: @test10(
; CHECK-NEXT:    [[T4:%.*]] = icmp eq ptr [[X:%.*]], [[Y:%.*]]
; CHECK-NEXT:    ret i1 [[T4]]
;
  %t1 = getelementptr { i32, i32 }, ptr %x, i32 0, i32 1
  %t3 = getelementptr { i32, i32 }, ptr %y, i32 0, i32 1
  %t4 = icmp eq ptr %t1, %t3
  ret i1 %t4
}
```

The crash report indicates assertion failure in `APInt.h` within `setBit()`. It appears to be caused by the fact that for `ptr` types, bit size is reported as 0, so later in the matching code, a call to `APInt::getOneBitSet` essentially asks for a number that has 0 bits, with bit #0 set. I'm unsure if I should simply ignore cases like this (e.g. by adding `if (Op0->getType()->getScalarSizeInBits() == 0) return nullptr`), or if there's a nicer/more usual way of handling this?

goldsteinn commented:
> Oops, looks like I forgot to run tests on my machine before pushing the changes. I've localised the source of one of many failing tests to this function in `getelementptr.ll`:
> 
> ```
> define i1 @test10(ptr %x, ptr %y) {
> ; CHECK-LABEL: @test10(
> ; CHECK-NEXT:    [[T4:%.*]] = icmp eq ptr [[X:%.*]], [[Y:%.*]]
> ; CHECK-NEXT:    ret i1 [[T4]]
> ;
>   %t1 = getelementptr { i32, i32 }, ptr %x, i32 0, i32 1
>   %t3 = getelementptr { i32, i32 }, ptr %y, i32 0, i32 1
>   %t4 = icmp eq ptr %t1, %t3
>   ret i1 %t4
> }
> ```
> 
> The crash report indicates assertion failure in `APInt.h` within `setBit()`. It appears to be caused by the fact that for `ptr` types, bit size is reported as 0, so later in the matching code, a call to `APInt::getOneBitSet` essentially asks for a number that has 0 bits, with bit #0 set. I'm unsure if I should simply ignore cases like this (e.g. by adding `if (Op0->getType()->getScalarSizeInBits() == 0) return nullptr`), or if there's a nicer/more usual way of handling this?

I'd say just handle the splat vector case for now. If #85676 gets in before this, you could use that API.

Poseydon42 commented:
> I'd say just handle the splat vector case for now. If #85676 gets in before this, you could use that API.
Done, should be passing tests now.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+  {
+    Value *Dividend;
+    const APInt *Divisor;
+    if (match(Op0, m_UDiv(m_Value(Dividend), m_APInt(Divisor))) &&
+        Op1 == Dividend && !Divisor->isZero() && !Divisor->isOne()) {
```
```suggestion
    if (match(Op0, m_UDiv(m_Specific(Op1), m_APInt(Divisor))) &&  Divisor->ugt(1)) {
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+    if (match(Op0, m_UDiv(m_Value(Dividend), m_APInt(Divisor))) &&
+        Op1 == Dividend && !Divisor->isZero() && !Divisor->isOne() &&
+        !ICmpInst::isUnsigned(Pred)) {
+      return new ICmpInst(ICmpInst::getSwappedPredicate(Pred), Dividend,
```
This code seems unreachable :(

dtcxzyw replied:
Do you mean `m_SDiv`?

Poseydon42 replied:
Yes, looks like I copy-pasted the `match` arguments from the code above and forgot to change it to `sdiv`.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+  {
+    Value *V;
+    const APInt *Shift;
+    if (match(Op0, m_LShr(m_Value(V), m_APInt(Shift))) && Op1 == V &&
+        !Shift->isZero()) {
```
```suggestion
    if (match(Op0, m_LShr(m_Specific(Op1), m_APInt(Shift))) &&
        !Shift->isZero()) {
```

goldsteinn commented on llvm/test/Transforms/InstCombine/icmp-div-constant.ll:
```diff
@@ -375,3 +375,115 @@ define i1 @sdiv_eq_smin_use(i32 %x, i32 %y) {
+;
+  %v = lshr i32 %x, 0
+  %r = icmp slt i32 %v, %x
+  ret i1 %r
```
to make any of these tests meaningful, they really need to be in the form of a non-splat vector constant where one element is the identity, and another is some interesting constant.

Poseydon42 replied:
The problem is, non-splat vectors won't be touched by my code either as they will fail the match due to the issue described above.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+    }
+  }
+
+  // Another case of this fold is (X >> Y) cmp X => X ~cmp 0 if Y != 0
```
nit: in your comments 'cmp' -> 'pred'.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+                          Constant::getNullValue(V->getType()));
+    }
+  }
+
```
Does it hold for `ashr`?
An example: https://alive2.llvm.org/ce/z/peP7WS

Poseydon42 replied:
Only with `slt` and `sge`, so I thought it not being worth implementing, as the potential gain seems minimal.

dtcxzyw replied:
It makes sense to me.

dtcxzyw reviewed (changes requested):
https://github.com/llvm/llvm-project/pull/79007/files may help :)

Poseydon42 commented:
Updated proofs with `ashr`: https://alive2.llvm.org/ce/z/52droC

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,37 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+                          Constant::getNullValue(Op1->getType()));
+    }
+
+    if (match(Op0, m_AShr(m_Specific(Op1), m_APInt(Shift))) && Shift->ugt(0) &&
```
nit: `!Shift->isZero()`

dtcxzyw replied:
It is not fixed.

Poseydon42 replied:
I believe it is fixed now.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,37 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+      return new ICmpInst(ICmpInst::getSwappedPredicate(Pred), Op1,
+                          Constant::getNullValue(Op1->getType()));
+    }
+  }
```
nit: There is a bit of redundancy in that each of the return cases is identical, but at least imo its fine.

goldsteinn commented:
LGTM, wait on one more approval.

Poseydon42 commented:
Ping @nikic @dtcxzyw

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,37 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+                          Constant::getNullValue(Op1->getType()));
+    }
+
+    if (match(Op0, m_SDiv(m_Specific(Op1), m_APInt(Divisor))) &&
+        Divisor->ugt(1) && !ICmpInst::isUnsigned(Pred)) {
```
```suggestion
    if (!ICmpInst::isUnsigned(Pred) && match(Op0, m_SDiv(m_Specific(Op1), m_APInt(Divisor))) &&
        Divisor->ugt(1)) {
```
`!ICmpInst::isUnsigned(Pred)` is cheaper.

Poseydon42 replied:
Fixed.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,39 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
   if (Value *V = foldICmpWithLowBitMaskedVal(Pred, Op0, Op1, Q, *this))
     return replaceInstUsesWith(CxtI, V);
 
+  // Folding (X / Y) pred X => X ~pred 0 for some constant Y other than 0 or 1
```
nit: `~pred` looks like the inverse predicate, not the swapped predicate, which are different things. I'd write this either as
```suggestion
  // Folding (X / Y) pred X => 0 pred X for some constant Y other than 0 or 1
```
or
```suggestion
  // Folding (X / Y) pred X => X swap(pred) 0 for some constant Y other than 0 or 1
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7103,6 +7103,39 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+  {
+    const APInt *Divisor;
+    if (match(Op0, m_UDiv(m_Specific(Op1), m_APInt(Divisor))) &&
+        Divisor->ugt(1)) {
```
Worth noting that this condition could be != 1 rather than > 1, because division by zero is UB anyway. Ultimately makes no difference in terms of behavior though, so your variant is fine as well.

nikic commented on llvm/test/Transforms/InstCombine/icmp-div-constant.ll:
```diff
@@ -375,3 +375,144 @@ define i1 @sdiv_eq_smin_use(i32 %x, i32 %y) {
+  ret <4 x i1> %r
+}
+
+define i1 @lshr_by_const_cmp_uge_value(i32 %x) {
```
```suggestion
define i1 @lshr_by_const_cmp_sge_value(i32 %x) {
```

nikic reviewed (approved):
LGTM

Poseydon42 commented:
Thanks nikic, looks like it is now ready for another review.
Ping @dtcxzyw
